src/game/boe.specials.cpp

```diff
@@ -2090,6 +2090,10 @@ cSpecial get_node(spec_num_t cur_spec, eSpecCtxType cur_spec_type) {
 			}
 			return univ.out->specials[cur_spec];
 		case eSpecCtxType::TOWN:
+			if (is_out()) {
+				showError("The scenario called a town special node but it is not in town.");
+				return dummy_node;
```

Comment:
What issue does this fix? I get that you're not supposed to call town specials while outdoors, but given that the commit message says "asan", I'm guessing you had a memory-related motive for enforcing it.

Is there any reason why the reverse is not also forbidden? You've forbidden calling a town special while outdoors, but not an outdoors special while in town, which seems odd.

Also, is there any chance that this causes special nodes in outdoors combat to fail? Testing outdoor encounters that call nodes on meet, win, or flee needs to be taken care of, obviously, but if I recall correctly, I also introduced a way to load an arbitrary town as your outdoor arena. I don't remember whether special encounters placed in that town will trigger, but if so, we would need to make sure that they work too.

Reply:
> What issue does this fix? I get that you're not supposed to call town specials while outdoors, but given that the commit message says "asan", I'm guessing you had a memory-related motive for enforcing it.
> 
> Is there any reason why the reverse is not also forbidden? You've forbidden calling a town special while outdoors, but not an outdoors special while in town, which seems odd.
> 
> Also, is there any chance that this causes special nodes in outdoors combat to fail? Testing outdoor encounters that call nodes on meet, win, or flee needs to be taken care of, obviously, but if I recall correctly, I also introduced a way to load an arbitrary town as your outdoor arena. I don't remember whether special encounters placed in that town will trigger, but if so, we would need to make sure that they work too.

This is from https://github.com/fosnola/cboe/commit/1e791f575e5dff039554ade1c89b390341a0e5f5 , so for some reason, get_node was called with a town special when I was out. But this is probably a quick fix for another problem that was fixed afterwards (maybe a city event that was stacked and called after calling end_town_mode).

We can probably keep this patch as a sanity check ( and add a similar patch when we are outdoor).

Reply:
Keeping this patch (and a similar patch for outdoors) is fine as long as it doesn't cause any problems with specials in outdoor combat.

Reply:
I am not sure that I understand. If we are in outdoor combat, overall_mode is equal to MODE_COMBAT ? So is_out() will return false:
```
bool is_out() {
        if((overall_mode == MODE_OUTDOORS) || (overall_mode == MODE_LOOK_OUTDOORS))
                return true;
        else if(overall_mode == MODE_SHOPPING) {
                std::swap(overall_mode, store_pre_shop_mode);
                bool ret = is_out();
                std::swap(overall_mode, store_pre_shop_mode);
                return ret;
        } else if(overall_mode == MODE_TALKING) {
                std::swap(overall_mode, store_pre_talk_mode);
                bool ret = is_out();
                std::swap(overall_mode, store_pre_talk_mode);
                return ret;
        } else return false;
}
```
Ie. I don't think it's possible to shop or talk when we're in external combat, but maybe I'm wrong...

Reply:
Ah, so I think that means there should be no issue either with this or with a similar check in the outdoors case. The on meet special is called before combat starts, and on flee/win is called after it ends.